### PR TITLE
Add multiple backend support to cinder_service_check.py

### PR DESCRIPTION
**Pull request by claco**
When the host argument is supplied and the host has multiple cinder
backends configured, duplicate status metrics are returned. MaaS
currently only checks the last value during alarm evaluation. If the
last backend metric is 1 (up) but previous backend metrics are 0 (down),
no MaaS alert will be fired.

This change addresses this by evaluating all services for a host first,
and then aggregating the results into one status metric.

Additionally, each backend on the host now has its own status metric to
aid in troubleshooting and allow for alarms to be set on individual
backends.

Issue #836

### CODE REVIEW

**Comment by claco** [comment]
Builds failed that maas check when the host returned in the services list has no @ at all (354d7f5e-block1), which wasn't a case I was able to test for in the heat stacks I'd run. Adjusting that `split('@')` and re pushing.

**Comment by claco** [comment]
Gate failures unrelated. Something is afoot with SQL-Alchemy in pip land:  http://lists.openstack.org/pipermail/openstack-dev/2016-February/087486.html

```
error in SQLAlchemy-Utils setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
```

**Comment by prometheanfire** [comment]
recheck - let's see if they fixed gate

**Comment by sigmavirus24** [comment]
I think setuptools released a fixed version. I saw a commit related to this fly by in IRC this morning. The bug was not in sqlalchemy-utils or openstack but in setuptools.

**Comment by sigmavirus24** [comment]
What actually changed is that (as Jason mentions here: https://bitbucket.org/pypa/setuptools/issues/502/packaging-164-does-not-allow-whitepace) he pulled the 20.2.1 release from PyPI. This means that we no longer pull it down.

**Comment by prometheanfire** [comment]
python looks fine, and AIO built/ran meaning maas was tested :+1:

**Comment by sigmavirus24** [comment]
> AIO built/ran meaning maas was tested

There's no test that these metrics were generated/reported to maas though. We'll need a real lab to do that that has cinder configured with *multiple* backends. Since this is meant to fix such a bug, I think relying on the AIO alone in this case is suboptimal.

**Comment by claco** [comment]
What's the recommendation for these types of things where's it's a customer edge case? @seancarlisle  might be able to test it for confirmation. Gating aside, I'm also more than happy to spin a setup w/ MaaS that has multiple backends for some eyeball verifications if that helps.

**Comment by prometheanfire** [comment]
it'd help for the feel goods I guess

**Comment by sigmavirus24** [comment]
We could also add a wishlist item to add actual unit testing to these scripts. I think spinning up labs and asking others to do verification for us is rather inefficient and a poor quality test. That said, I doubt we have time to write some automated testing of these plugins.

File: maas/plugins/cinder_service_check.py [hunk 1]
```diff
@@ -72,17 +72,32 @@ def check(auth_ref, args):
...
+            if service['status'] == 'enabled' and service['state'] != 'up':
+                service_is_up = False
+                all_services_are_up = False
+
+            if '@' in service['host']:
+                [host, backend] = service['host'].split('@')
+                name = '%s-%s_status' % (service['binary'], backend)
+                metric_bool(name, service_is_up)
+
+        name = '%s_status' % service['binary']
```
Comment (git-harry) [inline maas/plugins/cinder_service_check.py]: This is making an assumption that service['binary'] is the same across all services. This is probably a safe; the scheduler and volume services used to exist on the same host but that should no longer be the case. I just wanted to point it out incase anyone can think of potential issues.

**Comment by claco** [comment]
I'm going to stand up a stack today for @seancarlisle to take a gander at this change.

**Comment by sigmavirus24** [comment]
I was able to test this on @claco's stack and it works as expected. Just want to make sure that @seancarlisle is okay with the names that @claco is generating for the checks then I will merge-y merge-y

**Comment by prometheanfire** [comment]
@sigmavirus24 is that your :+1: on this?

**Comment by sigmavirus24** [comment]
@prometheanfire you replied to this after normal 9-5. ;)

That's a +0.5. I think @seancarlisle should contribute the other 0.5

**Comment by prometheanfire** [comment]
this is to master, therefore 12.1.0 right now

**Comment by seancarlisle** [comment]
The code looks alright. My only comment would be that the else section will never generate an alert because it will pass a metric for "cinder-volume_on_host_blah." Was this meant as more of a local test?

**Comment by claco** [comment]
@seancarlisle I guess that's the nebulous part. I don't think we never had any alarms on that no-host-passed scenario.

```
https://github.com/rcbops/rpc-openstack/blob/master/rpcd/playbooks/roles/rpc_maas/templates/cinder_volume_check.yaml.j2

https://github.com/rcbops/rpc-openstack/blob/master/rpcd/playbooks/roles/rpc_maas/templates/cinder_scheduler_check.yaml.j2
```

I'm pretty sure both of those pass host, and I don't see and alarms configured that don't pass host.

```
https://github.com/rcbops/rpc-openstack/search?utf8=%E2%9C%93&q=cinder_service_check.py
```

Good catch though.

**Comment by claco** [comment]
Just to clarify for those following, I forgot that the scheduler check uses the scheduler container ip as the host are, so the output is as expected. The alarm is still green.

image

**Comment by prometheanfire** [comment]
once it's confirmed this is intended behavior then merge

@BjoernT ?

**Comment by BjoernT** [comment]
:+1:

**Comment by sigmavirus24** [comment]
@BjoernT confirmed this independently.
